internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1072,123 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
 	}
 	return nil, nil
 }
+
+// checkOldSnapshotsPipelineRun sorts all snapshots for application and cancells all running interationTest pipelineruns within application
+// removes finalizer before the pipelinerun is set as cancelled run finally to be gracefuly cancelled
+// god please forgive me for this monstrosity of cognitive complexity 40+
+func (a *Adapter) checkAndCancelOldSnapshotsPipelineRun(application *applicationapiv1alpha1.Application, snapshot *applicationapiv1alpha1.Snapshot) error {
+
+	snapshots, err := a.loader.GetAllSnapshotsForGivenPR(a.context, a.client, application, snapshot.GetAnnotations()[gitops.PipelineAsCodePullRequestAnnotation])
+	if err != nil {
+		a.logger.Error(err, "Failed to fetch Snapshots for the application",
+			"application.Name:", application.Name)
+		return err
+	}
+	sortedSnapshots := gitops.SortSnapshots(*snapshots)
+	// check for older snapshots, their current status and type
+	latestSnapshot := sortedSnapshots[0]
+	for i := 1; i < len(sortedSnapshots); i++ {
+
+		if gitops.IsGroupSnapshot(&sortedSnapshots[i]) {
```

Comment:
Currently we don't have pr number in `group` snapshot because `group` snapshot can be created for one or multiple PRs, so the returned snapshot in `GetAllSnapshotsForGivenPR` won't include `group` snapshot

Reply:
Is there any other label for group snapshots, that we could use for determining which we can cancel?

Reply:
If you want to get the group snapshot created for that, I feel you may query with the label `test.appstudio.openshift.io/pr-group-sha`

Reply:
Repo url is not enough to differentiate Snapshots, we'd start cancelling tests for unrelated PRs.
I'd prefer not to cancel group Snapshots rather than doing that.

Reply:
I don't agree with that approach, getting Snapshots for a given repo is not nearly precise enough for us to decide to cancel running integration tests.
Let's say we open a PR with a change in the repo and our integration tests start running for an hour.
Then dependabot opens a new PR to bump the version of an image - this is a separate PR and shouldn't cause our running CI to be cancelled.

Reply:
Maybe it would be a good idea to have a follow-up story to cover group test Snapshots instead of trying to cover it in this change?

Reply:
Sounds right. Lets drop the group snapshots part for this change and create follow up story that will take care of that.

Reply:
Removed the group snapshots logic, together with unit tests, lets see what change this brings to e2e tests.